foundation/org.eclipse.persistence.core/src/org/eclipse/persistence/mappings/AggregateCollectionMapping.java

```diff
@@ -2526,31 +2526,18 @@ public void preInsert(WriteObjectQuery query) throws DatabaseException, Optimist
 
     /**
      * INTERNAL:
-     * Returns clone of InsertObjectQuery from the reference descriptor, if it is not set - create it.
+     * Returns a clone of InsertObjectQuery from the ClassDescriptor's DescriptorQueryManager or a new one
      */
     protected InsertObjectQuery getInsertObjectQuery(AbstractSession session, ClassDescriptor desc) {
-        DescriptorQueryManager queryManager = desc.getQueryManager();
-        if (!queryManager.hasInsertQuery()) {
-            synchronized (queryManager) {
-                if (!queryManager.hasInsertQuery()) {
-                    queryManager.setInsertQuery(new InsertObjectQuery());
-                }
-            }
-        }
-        InsertObjectQuery insertQuery = queryManager.getInsertQuery();
-        if (insertQuery.getModifyRow() == null) {
-            AbstractRecord modifyRow = new DatabaseRecord();
-            for (int i = 0; i < getTargetForeignKeyFields().size(); i++) {
-                DatabaseField field = getTargetForeignKeyFields().elementAt(i);
-                modifyRow.put(field, null);
-            }
-            desc.getObjectBuilder().buildTemplateInsertRow(session, modifyRow);
-            getContainerPolicy().addFieldsForMapKey(modifyRow);
-            if(this.listOrderField != null) {
-                modifyRow.put(this.listOrderField, null);
-            }
-            insertQuery.setModifyRow(modifyRow);
+        InsertObjectQuery insertQuery = desc.getQueryManager().getInsertQuery();
+        if (insertQuery == null) {
+            insertQuery = new InsertObjectQuery();
+            insertQuery.setDescriptor(desc);
```

Comment:
So, we have removed ```queryManager.setInsertQuery``` which makes the race condition moot. I'm not sure of the ramifications of this... but perhaps it's right.

Reply:
Another thing to point out is that as long as```desc.getQueryManager().getInsertQuery() == null```, we are going to keep making unique InsertObjectQuerys instance. These each will need to be prepared. The older implementation benefited from a single InsertObjectQuery instance that would only be prepared once.

Reply:
Agree with you here, on the other hand we gain correctness. I believe that we can gain some penalty introduced by this back by setting `insertQuery.setIsExecutionClone(true);` later to avoid unnecessary double cloning - one explicit, the other during execute

Reply:
I'm not sure I understand why we are gaining correctness. Previously, we were setting the InsertObjectQuery back on the QueryManager. That was where the issue originally was due to the timing of setting InsertObjectQuery multiple times (hence the synch block I added). Now, we are dropping that and instead mass producing cloned instances. 
This isn't to say I disagree with making the change, but that I cannot say I understand why this code is "correct" over the previous.

Reply:
By gaining correctness I mean:
* we're setting classdescriptor on the query instance before calling checkPrepare
* mapping is not altering QueryManager during execution; only initialization code is supposed to do that

Reply:
We are setting the classdescriptor on the query instance with the current implementation as well, yes? Is there a document somewhere that says we should not be altering QueryManager during execution? Perhaps if ```DescriptorQueryManager.setInsertQuery``` was documented to say that it should only be called during initialization. I have seen documentation like this on other methods.

Reply:
> We are setting the classdescriptor on the query instance with the current implementation as well, yes?

With the one in this PR - yes

>  Is there a document somewhere that says we should not be altering QueryManager during execution? 

I haven't found any. What I've found is general official documentation on the purpose of DescriptorQueryManager itself, ie `The DescriptorQueryManager of each Descriptor lets you customize the query implementation that EclipseLink uses for the data source operations.` - that says nothing about mappings. It even does not mention that those queries may be customized throught the mapping. But IMO it implies that mapping code _should not_ silently modify its state which is shared across whole application. Should the query need to be cached, mapping can do that in its own private/protected field and not necessarily through the manager